Blog - Jobs and Cron Jobs in Kubernetes (#473)
earthly/website: blog/_posts/2023-06-14-kubernetes-jobs-cron-jobs.md

---
title: "Jobs and Cron Jobs in Kubernetes"
categories:
  - Tutorials
toc: true
author: Mercy Bassey
editor: Bala Priya C

internal-links:
- Jobs
- Cron Jobs
- Kubernetes
- scheduled tasks
- Management
---

> 

Reviewer: ```suggestion

**We're [Earthly.dev](https://earthly.dev/). We make building software simpler and therefore faster using containerization. This article covers jobs and cron jobs in Kubernetes. If you want to know more about building in containers then [check us out](/).**
```

Modern software applications require efficient and reliable management of jobs and scheduled tasks. As applications become more complex and Kubernetes clusters increase in size, automating repetitive tasks like backups, batch processing, and data analysis becomes essential.

> Kubernetes provides two critical resources to manage these tasks: Jobs and CronJobs. Jobs are ideal for running one-off or batch tasks, while a CronJob is perfect for scheduling recurring tasks. By using these resources, developers, DevOps engineers, and cluster administrators can focus on more critical tasks and ensure the reliability and efficiency of Kubernetes deployments.

Reviewer: ```suggestion
Kubernetes provides two critical resources to manage these tasks: **Jobs** and **CronJobs**. Jobs are ideal for running one-off or batch tasks, while a CronJob is perfect for scheduling recurring tasks. By using these resources, developers, DevOps engineers, and cluster administrators can focus on more critical tasks and ensure the reliability and efficiency of Kubernetes deployments.
```

Whether you're a seasoned developer or a Kubernetes novice, understanding Jobs and CronJobs is essential for efficient and reliable application management. In this tutorial, we'll explore how to run and schedule tasks with Kubernetes Jobs and CronJobs. You'll learn how to create (define the desired state), monitor (watch their completion status), and customize these powerful resources in Kubernetes!

## Prerequisites

To follow along in this tutorial, you'll need to have:

- A Kubernetes cluster up and running (A single node cluster will work fine).

> - This tutorial uses a Linux machine with the Ubuntu 22.04LTS distro (Any other OS will work fine as long as you have Kubernetes set up).

Reviewer: ```suggestion
- This tutorial uses a Linux machine with the Ubuntu 22.04 LTS distro (Any other OS will work fine as long as you have [Kubernetes set up](/blog/k8s-dev-solutions/)).
```

### Creating a Job Manifest File

Before you can run a job in Kubernetes, you need to create a job manifest file that defines the specifications of the job. This file should be created in YAML format and should include information such as the name of the job, the container image to be used, and the command to run.

To create your first job, create a *yaml* file, open it up with your preferred code editor, and paste in the following code snippets; this tutorial uses the *[nano](https://docs.nano.org/)* text editor and calls this file `job.yaml`:

~~~{.yaml caption="job.yaml"}
# job.yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: hello-world
spec:
  template:
    spec:
      containers:
      - name: busybox
        image: busybox
        command: ["echo",  "Hello World!"]
      restartPolicy: Never
~~~

The code snippet above does the following:

- Creates a Kubernetes job named *hello-world* using the *batch/v1* API version.
- Runs a container named *busybox* with the image *busybox*.
- Executes the command *echo Hello World!* and then exits.

<div class="notice--info">

> 💡The *`restartPolicy: Never`* specifies that the container should not be restarted if it fails or exits.

Reviewer: ```suggestion
💡 The `restartPolicy: Never` specifies that the container should not be restarted if it fails or exits.
```
